1.21.5 component changes

Gerrygames opened the pull request:
things that need to be looked at:
- [x] show_entity field renames
- [x] change_page is now an int (or number?)
- [x] order of checks when deserializing hover events

Gerrygames committed:
1.21.5 component changes

Co-authored-by: MiniDigger | Martin <admin@minidigger.dev>

turikhay commented on text-serializer-gson/src/main/java/net/kyori/adventure/text/serializer/gson/StyleSerializer.java:
```diff
@@ -99,6 +110,9 @@ static TypeAdapter<Style> create(final net.kyori.adventure.text.serializer.json.
   private final net.kyori.adventure.text.serializer.json.LegacyHoverEventSerializer legacyHover;
   private final boolean emitLegacyHover;
   private final boolean emitModernHover;
+  private final boolean emitSuperModernHover;
```
I'm not sure if `superModern` is a right fit here. What if Mojang introduced further changes, would it be `superDuperModern`?

kezz replied:
Agreed. Let's deprecate the old options and be explicit in either what they do, or what version the changes were introduced in.

MiniDigger replied:
ideally end users should just use something like 
```java
GsonComponentSerializer.builder()
              .options(JSONOptions.byDataVersion().at(Bukkit.getUnsafe().getDataVersion()))
              .build();
```
anyways

MiniDigger replied:
pushed something, how does that feel?

turikhay replied:
It's way better

MiniDigger commented:
(rebased to fix conflicts)

MiniDigger commented on text-serializer-gson/src/main/java/net/kyori/adventure/text/serializer/gson/ShowEntitySerializer.java:
```diff
@@ -60,9 +65,10 @@ public HoverEvent.ShowEntity read(final JsonReader in) throws IOException {
     while (in.hasNext()) {
       final String fieldName = in.nextName();
-      if (fieldName.equals(SHOW_ENTITY_TYPE)) {
+      // TODO make this more flexible
```
I cant think of a way to make this support old and new rn, need input

turikhay replied:
In any case, I don't think this can be done in backwards-compatible way.

We should probably create a wrapper for `HoverEvent.ShowItem` and `HoverEvent.ShowEntity`, and serialize that instead. Wrapper would include all necessary context for Serializers to work with.

MiniDigger commented:
rebased on adventure 4.20 and fixed the failing tests

kezz committed:
peek to try and determine if we're looking at a key or uuid

this will only break if someone has a minecraft key with a valid uuid value and they don't provide the namespace... but like, meh?
